halo2_proofs/src/plonk/evaluation.rs

```diff
@@ -291,242 +405,582 @@ impl<C: CurveAffine> Evaluator<C> {
-            .iter()
-            .map(|instance_polys| {
-                instance_polys
-                    .iter()
-                    .map(|poly| domain.coeff_to_extended(poly.clone()))
-                    .collect()
-            })
-            .collect();
-        stop_measure(start);
+        assert!(self.custom_gate_clusters.len() <= num_clusters);
 
-        let mut values = domain.empty_extended();
+        // Initialize the the powers of y and constraint counter
+        let mut y_powers = vec![C::ScalarExt::ONE; self.num_y_powers * instance_polys.len()];
+        for i in 1..self.num_y_powers {
+            y_powers[i] = y_powers[i - 1] * y;
+        }
 
-        // Core expression evaluations
-        let num_threads = multicore::current_num_threads();
-        for (((advice, instance), lookups), permutation) in advice
-            .iter()
-            .zip(instance.iter())
-            .zip(lookups.iter())
-            .zip(permutations.iter())
+        let need_to_compute = |part_idx, cluster_idx| part_idx % (num_parts >> cluster_idx) == 0;
+        let compute_part_idx_in_cluster =
+            |part_idx, cluster_idx| part_idx >> (num_clusters - cluster_idx - 1);
+
+        let mut value_part_clusters = Vec::new();
+        value_part_clusters.resize(num_clusters, Vec::new());
+        for (cluster_idx, cluster) in value_part_clusters
+            .iter_mut()
+            .enumerate()
+            .take(num_clusters)
         {
-            // Custom gates
-            let start = start_measure("custom gates", false);
-            multicore::scope(|scope| {
-                let chunk_size = (size + num_threads - 1) / num_threads;
-                for (thread_idx, values) in values.chunks_mut(chunk_size).enumerate() {
-                    let start = thread_idx * chunk_size;
-                    scope.spawn(move |_| {
-                        let mut eval_data = self.custom_gates.instance();
-                        for (i, value) in values.iter_mut().enumerate() {
-                            let idx = start + i;
-                            *value = self.custom_gates.evaluate(
-                                &mut eval_data,
-                                fixed,
-                                advice,
-                                instance,
-                                challenges,
-                                &beta,
-                                &gamma,
-                                &theta,
-                                &y,
-                                value,
-                                idx,
-                                rot_scale,
-                                isize,
-                            );
+             cluster.resize(1 << cluster_idx, domain.empty_lagrange());
+        }
+
+        // Calculate the quotient polynomial for each part
+        let mut current_extended_omega = one;
+        for part_idx in 0..num_parts {
```

Comment:
The `evaluate_h` function is now almost 600 lines, with almost 500 lines spent within this for loop.

This will make keeping the state and flow in head hard when reading and refactoring the code later.

We likely want either now or in a later sprint a refactoring to factor things in different functions.

For example:
- `evaluate_custom_gates`
- `evaluate_permutations`
- `evaluate_lookups`
- ...